include/sys/semihost.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2022, Commonwealth Scientific and Industrial Research
+ * Organisation (CSIRO) ABN 41 687 119 230.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * https://developer.arm.com/documentation/dui0471/m/what-is-semihosting-
+ */
+
+/**
+ * @file
+ *
+ * @brief public ARM semihosting APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_SEMIHOST_H_
+#define ZEPHYR_INCLUDE_SYS_SEMIHOST_H_
+
+/**
+ * @brief Modes to open a file with
+ *
+ * Behaviour corresponds to equivalent fopen strings.
+ * i.e. SEMIHOST_OPEN_RB_PLUS == "rb+"
+ */
+enum semihost_open_mode {
+	SEMIHOST_OPEN_R         = 0,
+	SEMIHOST_OPEN_RB        = 1,
+	SEMIHOST_OPEN_R_PLUS    = 2,
+	SEMIHOST_OPEN_RB_PLUS   = 3,
+	SEMIHOST_OPEN_W         = 4,
+	SEMIHOST_OPEN_WB        = 5,
+	SEMIHOST_OPEN_W_PLUS    = 6,
+	SEMIHOST_OPEN_WB_PLUS   = 7,
+	SEMIHOST_OPEN_A         = 8,
+	SEMIHOST_OPEN_AB        = 9,
+	SEMIHOST_OPEN_A_PLUS    = 10,
+	SEMIHOST_OPEN_AB_PLUS   = 11,
+};
+
+/**
+ * @brief Read a byte from the console
+ *
+ * @return char byte read from the console.
+ */
+char semihost_poll_in(void);
+
+/**
+ * @brief Write a byte to the console
+ *
+ * @param c byte to write to console
+ */
+void semihost_poll_out(char c);
+
+/**
+ * @brief Open a file on the host system
+ *
+ * @param path file path to open. Can be absolute or relative to current
+ *             directory of the running process.
+ * @param mode value from @ref semihost_open_mode.
+ *
+ * @retval handle positive handle on success.
+ * @retval -1 on failure.
+ */
+int semihost_open(const char *path, int mode);
+
+/**
+ * @brief Close a file
+ *
+ * @param fd handle returned by @ref semihost_open.
+ *
+ * @retval 0 on success.
+ * @retval -1 on failure.
+ */
+int semihost_close(int fd);
+
+/**
+ * @brief Query the size of a file
+ *
+ * @param fd handle returned by @ref semihost_open.
+ *
+ * @retval positive file size on success.
+ * @retval -1 on failure.
+ */
+int semihost_flen(int fd);
+
+/**
+ * @brief Seeks to an absolute position in a file.
+ *
+ * @param fd handle returned by @ref semihost_open.
+ * @param offset offset from the start of the file in bytes.
+ *
+ * @retval 0 on success.
+ * @retval negative error code on failure.
```

Comment:
nit: format is `@retval VALUE DESCRIPTION`, so maybe use `@retval -errno Negative error...`

Reply:
Would that be host `errno` codes or Zephyr `errno` codes?